src/bin/wasmtime.rs

```diff
@@ -99,11 +99,194 @@ impl Wasmtime {
+            Err(_) => WhichCli::Unspecified,
+        })
+    }
+
+    pub fn main() -> Result<()> {
+        match which_cli()? {
+            // If the old or the new CLI is explicitly selected, then run that
+            // variant no questions asked.
+            WhichCli::New => return Wasmtime::parse().execute(),
+            WhichCli::Old => return try_parse_old().unwrap_or_else(|e| e.exit()).execute(),
+
+            // fall through below to run an unspecified version of the CLI
+            WhichCli::Unspecified => {}
+        }
+
+        // Here it's not specified which version of the CLI should be used, so
+        // both the old and the new CLI parsers are used and depending on the
+        // results an execution happens.
+        let new = Wasmtime::try_parse();
+        let old = try_parse_old();
+        match (new, old) {
+            // Both parsers succeeded. This means that it's likely that no
+            // options to configure execution, e.g. `--disable-logging`, were
+            // used in the old parser or the new.
+            //
+            // The result of parsing can still differ though. For example:
+            //
+            //      wasmtime foo.wasm --invoke foo
+            //
+            // would previously be parsed as passing `--invoke` to Wasmtime but
+            // the new parser instead parses that as passing the flag to
+            // `foo.wasm`.
+            //
+            // In this situation use the `DEFAULT_OLD_BEHAVIOR` constant to
+            // dictate which wins and additionally print a warning message.
+            (Ok(new), Ok(old)) => {
+                if new == old {
+                    return new.execute();
+                }
+                if DEFAULT_OLD_BEHAVIOR {
+                    eprintln!(
+                        "\
+warning: this CLI invocation of Wasmtime will be parsed differently in future
+         Wasmtime versions, but its interpretation is not changing yet -- see
+         this online issue for more information:
+         https://github.com/bytecodealliance/wasmtime/issues/7384
+
+         CLI parsing can be temporarily configured with an environment variable:
+
+         - WASMTIME_NEW_CLI=0 to silence this warning and not change behavior, or
+         - WASMTIME_NEW_CLI=1 to force using the new behavior\
+"
+                    );
+                    old.execute()
+                } else {
+                    // this error message is not statically reachable due to
+                    // `DEFAULT_OLD_BEHAVIOR=true` at this time, but when that
+                    // changes this should be updated to have a more accurate
+                    // set of text.
+                    assert!(false);
+                    eprintln!(
+                        "\
+warning: this CLI invocation of Wasmtime is parsed differently now than it was
+         prior, and the new semantics are going to be used instead of the old
```

Comment:
"going to be used" reads somewhat ambiguously to me: that could mean "in a few months when we flip a switch", or it could mean "in a few nanoseconds when I go ahead and call this function". Perhaps "and the new semantics are now used instead of the old semantics" (i.e., present rather than future tense)?

Reply:
If you're ok I'd prefer to leave precise wordsmithing of this message to later. This won't ever actually get executed due to the value of `DEFAULT_OLD_BEHAVIOR` and the `assert!(false)` prior.

I don't know how to best word this message because its hedging what's going to happen in the future which I think will shape what this should precisely say.

FWIW though "going to be used" is attempting to model what's actually about to happen, e.g. `new.execute()` instead of `old.execute()` below. This is trying to be a warning of "if you intended for the old behavior then things are gonna be broken now"

Reply:
Sure, we can come back to it later! My thoughts still apply; fwiw, data point of one person, "going to be used" confused me, so I think it should probably be reworded somehow by the time we use this error message if we're hoping to be accurate/unambiguous.